### PR TITLE
[DOC] Add developer documentation for adding tasks and definitions

### PR DESCRIPTION
**Pull request by 36000**
This is a rough draft for developer documentation for adding tasks and definitions

### CODE REVIEW

File: docs/source/developing/tasks.rst [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+created using a selection of tasks based on the configuration provided by
+the user. Each task file has a `get_{task filename}_plan` method which is
+called by the :class:`AFQ.api.AFQ` class (for example, in AFQ.tasks.data
+there is: :func:`AFQ.tasks.data.get_data_plan`). This method uses the user's
+configuration choices (which it takes as input) to construct a Pimms plan
+from the tasks in the file. In most cases, new tasks can simply be added to
+the list of tasks being passed to `with_name`. These tasks are always added
+to the plan; however they will only be called if the user needs them. In a few
+cases, additional logic is required, like when two different tasks compute the
+same output, or if the task is from an AFQ.definition.
```
Comment (arokem) [inline docs/source/developing/tasks.rst:29]: I don't understand the latter part of this sentence. Does this mean something like "if the task is inherited from `AFQ.Definition`"?

File: docs/source/developing/tasks.rst [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+The outputs of each task can be accessed by the user through the
+AFQ object, either as an attribute or using a `export_` method,
+such as `export_b0`. In each of the 5 tasks files, there is a variable called
+`outputs`. Add your task output(s) to that variable for them to be attached to
```
Comment (arokem) [inline docs/source/developing/tasks.rst:34]: Maybe add a link that shows this in one of these?

**Review by arokem — COMMENTED** [review]
Overall, looks great. That checklist in particular is very handy. I had a few comments and suggestions in various places.

**Comment by 36000** [comment]
Closes #696

File: docs/source/developing/definitions.rst [hunk 3]
```diff
@@ -0,0 +1,37 @@
+Adding Definitions 
+~~~~~~~~~~~~~~~~~~
+pyAFQ uses definitions to allow users to customize the mappings, masks,
+etc. used by the :class:`AFQ.api.AFQ` object. All definitions ultimately
+inherit from :class:`AFQ.definitions.utils.Definition`. For a given subject and
+session within the API, a definition is used to create a given mask, map, etc.
+Definitions must have three methods: `__init__`, `find_path`, and `get_for_subses`.
+The requirements of each of these methods are described below:
+
+- Definitions have an init method which the users uses to specify
```
[the remark this answers is not in the document]
Comment (36000) [reply docs/source/developing/definitions.rst:10]: It is actually the user who instantiates Definition-inherited classes, so the init method is how the user configures the given Definition-inherited instance.

File: docs/source/developing/definitions.rst [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  how they want the definition to behave. These init methods must be thoroughly
+  documented as they are what the user interacts with. The class must have attributes
+  of same name as the init args. This is important for reading/writing definitions
+  as arguments to config files. For scalar defintions, each class should have a
+  name parameter.
```
[the remark this answers is not in the document]
Comment (36000) [reply docs/source/developing/definitions.rst:15]: It's both,  I will add a link to an example.

File: docs/source/developing/definitions.rst [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  how they want the definition to behave. These init methods must be thoroughly
+  documented as they are what the user interacts with. The class must have attributes
+  of same name as the init args. This is important for reading/writing definitions
+  as arguments to config files. For scalar defintions, each class should have a
+  name parameter. 
+
+- The api calls find_path during the :class:`AFQ.api.AFQ` object initialization to
+  let the definition find relevant files for the given subject and session. All find_path
+  methods have the same input: bids_layout, from_path, subject, session. See
+  :func:`AFQ.definitions.mask.MaskFile.find_path` for a basic example.
```
[the remark this answers is not in the document]
Comment (36000) [reply docs/source/developing/definitions.rst:20]: You can click on source to see the code in the documentation. I will specify that.

File: docs/source/developing/tasks.rst [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+data, mapping, tractography, segmentation, and visualization (viz). If you want
+to add new functionality to the pyAFQ workflow, first decide which file to
+put it in. Try to keep similar functionality in the same file. All tasks are
+decorated with `@pimms.calc("output_name")`, where `output_name` is used as
+input to other tasks and can be accessed by the user through the
+:class:`AFQ.api.AFQ` object.
+
+When a user instantiates an :class:`AFQ.api.AFQ` object, a Pimms plan is
+created using a selection of tasks based on the configuration provided by
+the user. Each task file has a `get_{task filename}_plan` method which is
```
[the remark this answers is not in the document]
Comment (36000) [reply docs/source/developing/tasks.rst:21]: It shows up as : Each task file has a _get_{task filename}_plan_ method

File: docs/source/developing/tasks.rst (hunk 1 above)
Comment (36000) [reply docs/source/developing/tasks.rst:29]: I clarified it:  "or if the task being provided by a `Definition`-inherited object
(see the source of :func:`AFQ.tasks.mapping.get_mapping_plan` for an example
of both)."

File: docs/source/developing/tasks.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (36000) [reply docs/source/developing/tasks.rst:22]: Is there a better way to show this?

**Comment by 36000** [comment]
@arokem this is ready for another review

File: docs/source/developing/definitions.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (arokem) [reply docs/source/developing/definitions.rst:22]: Ah yes - I see the `func` directive now. This should work.

File: docs/source/developing/tasks.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (arokem) [reply docs/source/developing/tasks.rst:22]: I think this is fine, but might be confused by the GitHub markdown rendering 😄

[the remark this answers is not in the document]
Comment (arokem) [reply docs/source/developing/tasks.rst:22]: Let's merge this and then revisit as needed.

**Review by arokem — APPROVED** [review]

**Pull request merged**